PSA compliance tests suite

orenc17 opened the pull request:
### Description

Add PSA compliance tests suite. 
**Note:** this PR has been expanded to contain the entire suite.
This PR include tests for:
1. PSA internal-trusted-storage.
2. PSA Crypto
3. PSA Attestation

Relies on PRs
 #9708 (merged)
 #9795 (merged)
 #9668 (merged)
 #9822 (merged)
upcoming mbedTLS release
Note: this PR will not build without these PRs

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [X] Test update
    [ ] Breaking change

### Reviewers

<!-- 
    Optional
    Request additional reviewers with @username
-->

@alzix @jaypit02 @dreemkiller

jaypit02 commented on TESTS/psa-compliance/test_s001/main.c:
```diff
@@ -0,0 +1,50 @@
+
+const psa_api_t psa_api = {
+    .framework_version     = pal_ipc_framework_version,
+    .version               = pal_ipc_version,
```
These structures are common to all tests. Therefore, these can be moved to a common file. Moving to common file is more scalable to consume any addition/deletion of an element. 

Also you may want add copyright header to such files.

orenc17 replied:
I prefer to generate the structure for each test
That way it won't mistakenly compiled to an mbed-app and waste flash

orenc17 replied:
after a check I've moved the struct to the framework directory

jaypit02 commented on components/TARGET_PSA/spm/spm_client.h:
```diff
@@ -73,23 +73,23 @@ psa_handle_t psa_connect(uint32_t sid, uint32_t minor_version);
+    const psa_invec *in_vec,
     size_t in_len,
-    const psa_outvec_t *out_vec,
+    const psa_outvec *out_vec,
```
As per the latest PSA FF spec, out_vec parameter is no more "const". It is now:
```suggestion
    psa_outvec *out_vec,
```

jaypit02 commented on features/frameworks/TARGET_PSA/mbed_lib.json:
```diff
@@ -0,0 +1,26 @@
+{
+    "name": "psa-compliance",
+    "config": {
```
FYI- These macros definition will be available in pal_config.h in PSA compliance test suite release.  Once you have pal_config.h, this json may not be required.

orenc17 replied:
This is the "mbed way" for configuration
We could add platform specific configuration in the future through this file

jaypit02 commented on TESTS/psa-compliance/test_s001/test_s001.c:
```diff
@@ -0,0 +1,168 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+**/
+#define ITS_TEST
```
Is is possible to maintain test specific mbed_lib.json to pass ITS_TEST macro? 
Idea is to avoid test editing.

orenc17 replied:
No

jaypit02 reviewed (commented):
Added review comments

alzix reviewed (changes requested):
`test_s001` name does not describe it actually tests PSA ITS implementation.
please rename `TESTS/psa-compliance/test_s001`  to `TESTS/psa-compliance/psa-ist-s001` and fix the importer

alzix commented:
@orenc17 - ping

cmonr commented:
Making a note here. I think this now relies on https://github.com/ARMmbed/mbed-os/pull/9192 instead.

mikisch81 commented:
@cmonr actually like many others, it depends on #9529

0xc0170 commented:
@orenc17 What is the status for this PR? The dependencies were integrated, weren't they?

orenc17 commented:
the porting is being continued by another team.. i believe they use this PR as a base

cmonr commented:
@orenc17 They'll still hit the same problem once they introduce a PR.

Commit:
Delete unneeded .mk files after test import

delete unneeded .mk files imported as part of the PSA compliance test suite

Commit:
add imported PSA compliance tests to astyle ignore list

Commit:
changes to test infrastructure to adapt it to green-tea

changes to test infrastructure to adapt it to green-tea
Make ITS testing default
Fix IAR build issues

Commit:
[api-tests]: Updated to cb58d2c

Commit:
Allow PS test03 with PS_ALLOW_ENTIRE_STORAGE_FILL flag

Commit:
add main functions to each compliance test

add main function to each compliance test , moving shared code to a common file

Commit:
rework for mbed-os intf for ACK tests

-rework test wrapper functions to simplify test main
-move greentea init before test setup code to avoid boards losing sync.

Commit:
Define psa_storage_info_t for backward compatibility

Commit:
Fix for resource leak in crypto compliance test 43

Destroy key after import key in a loop

Commit:
update importer hashes for latest fixes

Commit:
use common flag to allow test 3 for PS and ITS

add a compilation flag to allow tests which fill up the entire device storage.
storage test 003 (for both PS and its) is unsuitable for the regular CI because:
1. it takes a long time causing timeouts
2. it places a very high attrition load on the storage of the CI boards
    and this can cause them to fail much sooner.

Commit:
update importer hashes for latest changes

cmonr commented on features/frameworks/TARGET_PSA/pal/pal_mbed_os_crypto.cpp:
```diff
@@ -19,6 +21,42 @@ extern "C" psa_status_t psa_ps_reset();
+{
+#ifndef PSA_ATTESTATION_DISABLED
+    const uint8_t private_key_data[] = {
+       0x49, 0xc9, 0xa8, 0xc1, 0x8c, 0x4b, 0x88, 0x56,
```
~Where does this key come from, and how was it generated?~

NirSonnenschein replied:
This is a hard-coded attestation key used for testing the attestation feature. it was randomly generated.
The specific key chosen shouldn't matter to the test it just needs a key to be injected before it is run (in practice each decide is expected to have it's own randomly generated key).

cmonr commented on features/frameworks/TARGET_PSA/val_greentea.cpp:
```diff
@@ -0,0 +1,324 @@
+    if (IS_TEST_FAIL(status) || IS_TEST_SKIP(status))
+    {
+        GREENTEA_TESTSUITE_RESULT(false);
+        return;
```
Seems odd that this was needed.

NirSonnenschein replied:
if you mean the explicit return, you are right, it should not be needed.
It is likely an artifact from the previous implementation of the function. - Fixed

cmonr commented on features/frameworks/TARGET_PSA/val_greentea.cpp:
```diff
@@ -0,0 +1,324 @@
+    bool continue_test = true;
+
+    test_info.test_num = test_num;
+    if (boot.state == BOOT_NOT_EXPECTED || boot.state == BOOT_EXPECTED_CRYPTO)
```
I'm probably missing something, but how would this ever not be true?

`boot_t boot` isn't static.

NirSonnenschein replied:
good point this check should be removed - fixed

cmonr commented on features/frameworks/TARGET_PSA/val_greentea.cpp:
```diff
@@ -0,0 +1,324 @@
+        mbed_val_print(PRINT_ERROR, "Call to dispatch SF failed. Status=%x\n", status_of_call);
+    }
+
+    pal_ipc_close(*handle);
```
Why should the handle be closed here instead of outside of the function?

NirSonnenschein replied:
this function is called from a pointer to function in the original attestation test framework (prior to our adaptation to greentea) in the struct val_api_t. The original implementation had the calling semantic that this function frees the handle inside and we preserved this for future compatibility (had we implemented the test framework we would have done many things very differently).

cmonr commented on features/frameworks/TARGET_PSA/val_greentea.cpp:
```diff
@@ -0,0 +1,324 @@
+}
+
+
+}
```
```
} // extern "C"
```

NirSonnenschein replied:
good point, fixed

cmonr commented on features/frameworks/TARGET_PSA/pal/pal_mbed_os_crypto.cpp:
```diff
@@ -0,0 +1,69 @@
+#endif
+
+#if defined(MBEDTLS_ENTROPY_NV_SEED) || defined(COMPONENT_PSA_SRV_IPC)
+        inject_entropy();
```
Seems weird that this and the following line of code are indented.

NirSonnenschein replied:
you are right, however this file has since been replaced (git move) with the file pal_mbed_os_intf.cpp which has been refactored and no longer contains this anomaly

NirSonnenschein reviewed (commented):
Thanks @cmonr

Commit:
removed redundant code + added comment

Thanks @cmonr.
